source/non_matching/fe_immersed_values.cc

```diff
@@ -0,0 +1,184 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2021 - 2021 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/grid/tria_iterator.h>
+
+#include <deal.II/non_matching/fe_immersed_values.h>
+
+
+DEAL_II_NAMESPACE_OPEN
+namespace NonMatching
+{
+  template <int dim>
+  FEImmersedSurfaceValues<dim>::FEImmersedSurfaceValues(
+    const Mapping<dim> &                  mapping,
+    const FiniteElement<dim> &            element,
+    const ImmersedSurfaceQuadrature<dim> &quadrature,
+    const UpdateFlags                     update_flags)
+    : FEValuesBase<dim, dim>(quadrature.size(),
+                             element.dofs_per_cell,
+                             update_default,
+                             mapping,
+                             element)
+    , quadrature(quadrature)
+  {
+    initialize(update_flags);
+  }
+
+
+
+  template <int dim>
+  template <bool level_dof_access>
+  void
+  FEImmersedSurfaceValues<dim>::reinit(
+    const TriaIterator<DoFCellAccessor<dim, dim, level_dof_access>> &cell)
+  {
+    // Assert that the finite elements passed to the constructor and used by the
+    // DoFHandler used by this cell, are the same
+    Assert(static_cast<const FiniteElementData<dim> &>(*this->fe) ==
+             static_cast<const FiniteElementData<dim> &>(cell->get_fe()),
+           (typename FEValuesBase<dim>::ExcFEDontMatch()));
+
+    this->maybe_invalidate_previous_present_cell(cell);
+    this->present_cell = {cell};
```

Comment:
Does this call via the constructor https://github.com/dealii/dealii/blob/2890a38bc0909248658137c26984c6d8e9f276f8/include/deal.II/fe/fe_values.h#L3833-L3835 work for all compilers? If I remember correctly we've had trouble with some code like this for some of the compilers (but I believe the CI should reveal it if problematic).

Reply:
I don't know. This is the same way `present_cell` is reset in the other `FEValues` class, but this was recently rewritten. What does @peterrum think?

Reply:
Yes. This is a plain copy from the the other `FEValues` classes. I think we use `={}` at other places in the library. I think there are even compilers that complain if you are explicitly write the type twice...

Reply:
Well, since we've had a round of testers on it since it was merged, I think we are good. I just wanted to check because I've been bitten by these before.